src/ray/object_manager/pull_manager.cc

```diff
@@ -0,0 +1,303 @@
+      total_failed_chunk_reads_(0),
+      client_id_(client_id),
+      gen_(std::chrono::high_resolution_clock::now().time_since_epoch().count()) {}
+
+void PullManager::ReceivePushRequest(const UniqueID &push_id, const ObjectID &object_id,
+                                     const ClientID &client_id, int64_t chunk_index,
+                                     int64_t num_chunks,
+                                     std::vector<ClientID> *clients_to_cancel,
+                                     bool *start_timer) {
+  *start_timer = false;
+  RAY_CHECK(client_id != client_id_);
+  *clients_to_cancel = std::vector<ClientID>();
+
+  auto it = pulls_.find(object_id);
+  if (it == pulls_.end()) {
+    *start_timer = true;
+    auto insertion_it = pulls_.insert(
+        std::make_pair(object_id, std::unique_ptr<PullInfo>(new PullInfo(false))));
+
+    RAY_CHECK(insertion_it.second);
+    it = insertion_it.first;
+
+    auto &pull_info = it->second;
+    RAY_CHECK(!pull_info->required);
+    pull_info->push_id = push_id;
+    pull_info->client_receiving_from = client_id;
+    pull_info->InitializeChunksIfNecessary(num_chunks);
+  } else {
+    auto &pull_info = it->second;
+    if (pull_info->client_receiving_from.IsNil()) {
+      pull_info->push_id = push_id;
+      pull_info->client_receiving_from = client_id;
+      pull_info->InitializeChunksIfNecessary(num_chunks);
+      // Cancel it from all remote object managers that we've requested it from
+      // except for this one.
+      for (auto const &client_requested_from : pull_info->clients_requested_from) {
+        if (client_requested_from != client_id) {
+          clients_to_cancel->push_back(client_requested_from);
+        }
+      }
+      pull_info->clients_requested_from.clear();
+    } else {
+      // We are already receiving the object from some other remote object
+      // manager.
+      RAY_CHECK(pull_info->clients_requested_from.empty());
+    }
+  }
+
+  auto &pull_info = it->second;
+  RAY_CHECK(pull_info->num_in_progress_chunk_ids >= 0);
+  pull_info->num_in_progress_chunk_ids++;
+}
+
+// If the object is required and if we are not already receiving the object from
+// some object manager, request the object from any object managers that we have
+// not already requested the object from.
+void PullManager::NewObjectLocations(
+    const ObjectID &object_id, const std::unordered_set<ClientID> &clients_with_object,
+    std::vector<ClientID> *clients_to_request, bool *restart_timer) {
+  *clients_to_request = std::vector<ClientID>();
+  *restart_timer = false;
+
+  auto it = pulls_.find(object_id);
+  if (it == pulls_.end()) {
+    return;
+  }
+
+  auto &pull_info = it->second;
+  pull_info->clients_with_object = clients_with_object;
+
+  if (!pull_info->required) {
+    return;
+  }
+
+  // If we are already receiving the object, don't request it from any more
+  // object managers.
+  if (!pull_info->client_receiving_from.IsNil()) {
```

Comment:
what would happen if the remote node manager crashes when we're still pulling from it?

Reply:
That should be caught when the timer is expired and find new object managers to pull from when `PullManagers::TimerExpired()` is called

https://github.com/williamma12/ray/blob/9ff30b608ca7cdf9c4f8b8c27bf71e43f4a29ca2/src/ray/object_manager/object_manager.cc#L210-L216